editing/edit-with-branch-versioning/src/main/java/com/esri/samples/edit_with_branch_versioning/EditWithBranchVersioningController.java

```diff
@@ -0,0 +1,356 @@
+        switchVersion(userCreatedVersion);
+
+        // hide the form from the UI as the sample only allows 1 version to be created
+        createVersionVBox.setVisible(false);
+        switchVersionButton.setDisable(false);
+
+      } catch (Exception ex) {
+        // if there is an error creating a new version, display an alert and reset the UI
+        if (ex.getCause().toString().contains("The version already exists")) {
+          new Alert(Alert.AlertType.ERROR, "A version with this name already exists.\nPlease enter a unique name").show();
+        } else {
+          new Alert(Alert.AlertType.ERROR, "Error creating new version").show();
+        }
+        createVersionButton.setText("Create version");
+        createVersionButton.setDisable(false);
+      }
+    });
+  }
+
+  /**
+   * Apply local edits to the service geodatabase and switch branch version.
+   */
+  @FXML
+  private void handleSwitchVersionButtonClicked() {
+
+    if (serviceGeodatabase.getVersionName().equals(defaultVersion)) {
+      // if the current version is the default version, switch to the user created version
+      switchVersion(userCreatedVersion);
+    } else if (serviceGeodatabase.getVersionName().equals(userCreatedVersion)) {
+      // if the current version is the user created version, check if there are local edits
+      if (!serviceGeodatabase.hasLocalEdits()) {
+        // if there are no local edits switch to the default version
+        switchVersion(defaultVersion);
+      } else {
+        // if local edits exist apply the edits
+        ListenableFuture<List<FeatureTableEditResult>> resultOfApplyEdits = serviceGeodatabase.applyEditsAsync();
+        resultOfApplyEdits.addDoneListener(() -> {
+          try {
+            // check if the server edit was successful
+            List<FeatureTableEditResult> edits = resultOfApplyEdits.get();
+            if (edits == null || edits.isEmpty()) {
+              new Alert(Alert.AlertType.ERROR, "Error applying edits on server").show();
+            } else {
+              // if the edits were successful, switch to the default version
+              switchVersion(defaultVersion);
+            }
+          } catch (InterruptedException | ExecutionException e) {
+            new Alert(Alert.AlertType.ERROR, "Error applying edits on server").show();
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * Switch the active branch version.
+   *
+   * @param versionName name of the version to switch to
+   */
+  private void switchVersion(String versionName) {
+    ListenableFuture<Void> switchVersionResult = serviceGeodatabase.switchVersionAsync(versionName);
+    switchVersionResult.addDoneListener(() -> {
+      // check if the active version has switched successfully and update the UI
+      if (serviceGeodatabase.getVersionName().equals(versionName)) {
+        currentVersionLabel.setText("Current version: " + serviceGeodatabase.getVersionName());
```

Comment:
I've noticed there are a lot of calls to `serviceGeodatabase.getVersionName()`. It looks as if this is the central place where a change to that value happens, meaning perhaps it might be nice to save it as a member variable (e.g. `String currentGeodatabaseVersion`), from which other methods could read, avoiding the repeated calls of the methos .getVersionName.

However there's an argument to be made about that approach creating two 'sources of truth', and in a more complex app the actual `serviceGeodatabase` and a string `currentGeodatabaseVersion` could go out of sync, creating a whole other bunch of problems...

```suggestion
		serviceGeodatabaseVersion
        currentVersionLabel.setText("Current version: " + currentGeodatabaseVersion);
```

I'm not too attached to any solution, so I'll just leave this here as a comment rather than a request for a change 😛

Reply:
Yeah this is a really good point. I had this discussion with myself and I had the fear of ending up with two sources of truth which is why I stuck with the 'gets'. I'll leave this comment open for second review too.

Reply:
@Rachael-E

Reply:
@jenmerritt I'm ok with this the way it is, with the calls to `getVersionName`. I think it's readable and helps understand the flow: at least for me!